### PR TITLE
Use ARM runners to build the linux/arm64 docker image

### PR DESCRIPTION
**Pull request by enescakir**
### Extract sha and latest tags automatically with docker/metadata-action

Currently, I manually release our docker image and specify image tags. While working on building our docker images on multiple runners, I revisited the "docker/metadata-action" documentation. https://github.com/docker/metadata-action#tags-input. It can automatically extract short SHA and latest tags.

When I initiate the build, it generates images with the following tags: `<BRANCH_NAME>`, `<SHORT_SHA>`, and `latest` if it's the default branch.

### Separate docker build and push operations into different runners

We build and push docker images for both amd64 and arm64 architectures on a single runner machine. Due to the slow build of arm64 images on an x64 runner, we intend to use separate runners for each architecture.

The main challenge lies in building docker images on different runners and pushing them to the registry using the same tag but different architectures.

To address this, I followed the [Distribute build across multiple runners](https://docs.docker.com/build/ci/github-actions/multi-platform/#distribute-build-across-multiple-runners) example, which allows us to build images on multiple runners and publish them after a merge.

`docker` jobs build images digests and upload them as artifacts. `merge` job merges digests and push them to Docker Hub.

### Build arm64 docker image on an ARM runner

ARM-based machines have recently gained popularity. Thanks to Apple's M and AWS Graviton processors, the amd64 image is no longer enough. An arm64 docker image is now essential for popular open-source projects. We already build arm64 docker images using docker buildx with QEMU emulation. However, this process is quite slow, taking ~20 minutes to build the ARM image with emulation. In contrast, building our image on a machine with the same architecture typically takes about 2 minutes.

The previous commit allows us to build images on the different runners. `merge` job merges them to the single tag.

Fixes https://github.com/ubicloud/ubicloud/issues/138

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -46,13 +55,73 @@ jobs:
...
+        if: ${{ inputs.push_image }}
+        uses: actions/upload-artifact@v3
+        with:
+          name: digests
+          path: /tmp/digests/*
+          if-no-files-found: error
+          retention-days: 1
+
+  merge:
+    if: ${{ inputs.push_image }}
```
Comment (fdr) [inline .github/workflows/build.yml:85]: I'm a bit confused, I expect to see a greater percentage of the lines here to be about arm64, but it seems like a lot of it is about digesting?

Comment (enescakir) [reply .github/workflows/build.yml:85]: I divided it to 2 commits

Comment (enescakir) [reply .github/workflows/build.yml:85]: @fdr Any comment on that?

**Review by fdr — APPROVED** [review]
well that intermediate commit is very helpful (description and cross-reference both), and then the arm64 patch ends up looking like what I might have expected.

**Pull request merged**